File: src/context/providers.py
# -*- coding: utf-8 -*-
"""
Context Providers - 上下文提供者

实现动态上下文注入系统，从多个来源聚合上下文信息。
支持记忆系统、技能系统、项目配置等上下文源。
"""

import logging
from abc import ABC, abstractmethod
from dataclasses import dataclass, field
from enum import IntEnum
from pathlib import Path
from typing import List, Dict, Any, Optional, Protocol, runtime_checkable
import json

logger = logging.getLogger("niko-context")


class ContextPriority(IntEnum):
    """上下文优先级（数值越小优先级越高）"""
    CRITICAL = 0      # 必须包含
    HIGH = 10         # 高优先级
    NORMAL = 50       # 正常优先级
    LOW = 100         # 低优先级
    OPTIONAL = 200    # 可选


@dataclass
class ContextItem:
    """上下文项"""
    key: str
    value: Any
    source: str  # 来源标识
    priority: ContextPriority = ContextPriority.NORMAL
    metadata: Dict[str, Any] = field(default_factory=dict)
    token_estimate: int = 0  # 预估令牌数

    def to_prompt_segment(self) -> str:
        """转换为 prompt 片段"""
        if isinstance(self.value, dict):
            content = json.dumps(self.value, ensure_ascii=False, indent=2)
        elif isinstance(self.value, list):
            content = "\n".join(str(v) for v in self.value)
        else:
            content = str(self.value)

        return f"[{self.key}]\n{content}\n[/{self.key}]"


@runtime_checkable
class IContextProvider(Protocol):
    """上下文提供者协议"""

    @property
    def name(self) -> str:
        """提供者名称"""
        ...

    @property
    def priority(self) -> ContextPriority:
        """默认优先级"""
        ...

    async def get_context(
        self,
        query: Optional[str] = None,
        **kwargs
    ) -> List[ContextItem]:
        """
        获取上下文项

        Args:
            query: 查询字符串（用于相关性筛选）
            **kwargs: 额外参数

        Returns:
            上下文项列表
        """
        ...


class BaseContextProvider(ABC):
    """上下文提供者基类"""

    def __init__(self, name: str, priority: ContextPriority = ContextPriority.NORMAL):
        self._name = name
        self._priority = priority

    @property
    def name(self) -> str:
        return self._name

    @property
    def priority(self) -> ContextPriority:
        return self._priority

    @abstractmethod
    async def get_context(
        self,
        query: Optional[str] = None,
        **kwargs
    ) -> List[ContextItem]:
        """获取上下文项"""
        pass

    def _estimate_tokens(self, text: str) -> int:
        """估算令牌数（简化实现）"""
        # 粗略估计：英文约 4 字符/token，中文约 1.5 字符/token
        chinese_chars = sum(1 for c in text if '\u4e00' <= c <= '\u9fff')
        other_chars = len(text) - chinese_chars
        return int(chinese_chars / 1.5 + other_chars / 4)


class MemoryContextProvider(BaseContextProvider):
    """
    记忆上下文提供者

    从 UnifiedMemoryEngine 获取相关记忆作为上下文。

    使用示例:
        provider = MemoryContextProvider(memory_engine)
        items = await provider.get_context(query="角色背景")
    """

    def __init__(
        self,
        memory_engine: Optional[Any] = None,
        max_items: int = 10,
        relevance_threshold: float = 0.5
    ):
        super().__init__("memory", ContextPriority.HIGH)
        self._memory_engine = memory_engine
        self._max_items = max_items
        self._relevance_threshold = relevance_threshold

    async def get_context(
        self,
        query: Optional[str] = None,
        session_id: Optional[str] = None,
        memory_types: Optional[List[str]] = None,
        **kwargs
    ) -> List[ContextItem]:
        """
        获取记忆上下文

        Args:
            query: 搜索查询
            session_id: 会话 ID
            memory_types: 记忆类型过滤
        """
        if self._memory_engine is None:
            logger.warning("Memory engine not configured")
            return []

        items: List[ContextItem] = []

        try:
            # 尝试调用记忆引擎的搜索方法
            if hasattr(self._memory_engine, 'search'):
                results = await self._memory_engine.search(
                    query=query or "",
                    limit=self._max_items,
                    session_id=session_id
                )

                for result in results:
                    if hasattr(result, 'score') and result.score < self._relevance_threshold:
                        continue

                    content = getattr(result, 'content', str(result))
                    items.append(ContextItem(
                        key=f"memory_{len(items)}",
                        value=content,
                        source=self.name,
                        priority=self.priority,
                        metadata={
                            "score": getattr(result, 'score', 1.0),
                            "type": getattr(result, 'memory_type', 'unknown')
                        },
                        token_estimate=self._estimate_tokens(content)
                    ))

            # 获取会话上下文
            if session_id and hasattr(self._memory_engine, 'get_session_context'):
                session_ctx = await self._memory_engine.get_session_context(session_id)
                if session_ctx:
                    items.append(ContextItem(
                        key="session_context",
                        value=session_ctx,
                        source=self.name,
                        priority=ContextPriority.CRITICAL,
                        token_estimate=self._estimate_tokens(str(session_ctx))
                    ))

        except Exception as e:
            logger.error(f"Failed to get memory context: {e}")

        return items


class SkillContextProvider(BaseContextProvider):
    """
    技能上下文提供者

    从 SkillLoader 获取技能包内容作为上下文。

    使用示例:
        provider = SkillContextProvider(skill_loader)
        items = await provider.get_context(skill_ids=["fictional-dream"])
    """

    def __init__(
        self,
        skill_loader: Optional[Any] = None,
        max_skill_length: int = 4000
    ):
        super().__init__("skill", ContextPriority.NORMAL)
        self._skill_loader = skill_loader
        self._max_skill_length = max_skill_length

    async def get_context(
        self,
        query: Optional[str] = None,
        skill_ids: Optional[List[str]] = None,
        include_summary: bool = True,
        **kwargs
    ) -> List[ContextItem]:
        """
        获取技能上下文

        Args:
            query: 用于匹配技能的查询
            skill_ids: 指定的技能 ID 列表
            include_summary: 是否包含技能摘要
        """
        if self._skill_loader is None:
            logger.warning("Skill loader not configured")
            return []

        items: List[ContextItem] = []

        try:
            # 加载指定技能
            if skill_ids:
                for skill_id in skill_ids:
                    try:
                        content = self._skill_loader.load(skill_id)
                        # 截断过长内容
                        if len(content) > self._max_skill_length:
                            content = content[:self._max_skill_length] + "\n... (内容截断)"

                        items.append(ContextItem(
                            key=f"skill_{skill_id}",
                            value=content,
                            source=self.name,
                            priority=self.priority,
                            metadata={"skill_id": skill_id},
                            token_estimate=self._estimate_tokens(content)
                        ))
                    except FileNotFoundError:
                        logger.warning(f"Skill not found: {skill_id}")

            # 从查询解析 @skill 引用
            if query and hasattr(self._skill_loader, 'extract_refs'):
                refs = self._skill_loader.extract_refs(query)
                for skill_id in refs:
                    if skill_ids and skill_id in skill_ids:
                        continue  # 已加载
                    try:
                        content = self._skill_loader.load(skill_id)
                        if len(content) > self._max_skill_length:
                            content = content[:self._max_skill_length] + "\n... (内容截断)"

                        items.append(ContextItem(
                            key=f"skill_{skill_id}",
                            value=content,
                            source=self.name,
                            priority=self.priority,
                            metadata={"skill_id": skill_id, "from_ref": True},
                            token_estimate=self._estimate_tokens(content)
                        ))
                    except FileNotFoundError:
                        pass

            # 包含技能摘要
            if include_summary and hasattr(self._skill_loader, 'get_summary'):
                summary = self._skill_loader.get_summary()
                items.append(ContextItem(
                    key="available_skills",
                    value=summary,
                    source=self.name,
                    priority=ContextPriority.LOW,
                    token_estimate=self._estimate_tokens(summary)
                ))

        except Exception as e:
            logger.error(f"Failed to get skill context: {e}")

        return items


class ProjectContextProvider(BaseContextProvider):
    """
    项目上下文提供者

    从 .niko/ 目录读取项目配置和上下文。

    使用示例:
        provider = ProjectContextProvider(project_root="/path/to/project")
        items = await provider.get_context()
    """

    def __init__(
        self,
        project_root: Optional[str] = None
    ):
        super().__init__("project", ContextPriority.HIGH)
        self._project_root = Path(project_root) if project_root else Path.cwd()
        self._niko_dir = self._project_root / ".niko"

    async def get_context(
        self,
        query: Optional[str] = None,
        include_characters: bool = True,
        include_world: bool = True,
        include_outline: bool = True,
        **kwargs
    ) -> List[ContextItem]:
        """
        获取项目上下文

        Args:
            query: 查询字符串（用于相关性筛选）
            include_characters: 包含角色信息
            include_world: 包含世界观信息
            include_outline: 包含大纲信息
        """
        items: List[ContextItem] = []

        if not self._niko_dir.exists():
            logger.debug(f"No .niko directory found at {self._niko_dir}")
            return items

        try:
            # 读取项目配置
            config_file = self._niko_dir / "config.json"
            if config_file.exists():
                config = json.loads(config_file.read_text(encoding="utf-8"))
                items.append(ContextItem(
                    key="project_config",
                    value=config,
                    source=self.name,
                    priority=ContextPriority.HIGH,
                    token_estimate=self._estimate_tokens(str(config))
                ))

            # 读取角色信息
            if include_characters:
                chars_dir = self._niko_dir / "characters"
                if chars_dir.exists():
                    characters = []
                    for char_file in chars_dir.glob("*.json"):
                        try:
                            char_data = json.loads(char_file.read_text(encoding="utf-8"))
                            characters.append(char_data)
                        except json.JSONDecodeError:
                            pass

                    if characters:
                        items.append(ContextItem(
                            key="characters",
                            value=characters,
                            source=self.name,
                            priority=ContextPriority.NORMAL,
                            metadata={"count": len(characters)},
                            token_estimate=self._estimate_tokens(str(characters))
                        ))

            # 读取世界观
            if include_world:
                world_file = self._niko_dir / "world.json"
                if world_file.exists():
                    world = json.loads(world_file.read_text(encoding="utf-8"))
                    items.append(ContextItem(
                        key="world",
                        value=world,
                        source=self.name,
                        priority=ContextPriority.NORMAL,
                        token_estimate=self._estimate_tokens(str(world))
                    ))

            # 读取大纲
            if include_outline:
                outline_file = self._niko_dir / "outline.json"
                if outline_file.exists():
                    outline = json.loads(outline_file.read_text(encoding="utf-8"))
                    items.append(ContextItem(
                        key="outline",
                        value=outline,
                        source=self.name,
                        priority=ContextPriority.HIGH,
                        token_estimate=self._estimate_tokens(str(outline))
                    ))

        except Exception as e:
            logger.error(f"Failed to get project context: {e}")

        return items


class ContextAggregator:
    """
    上下文聚合器

    从多个提供者收集上下文并进行优先级排序和令牌预算管理。

    使用示例:
        aggregator = ContextAggregator()
        aggregator.add_provider(MemoryContextProvider(memory_engine))
        aggregator.add_provider(SkillContextProvider(skill_loader))
        aggregator.add_provider(ProjectContextProvider())

        # 获取聚合上下文
        items = await aggregator.get_context(
            query="角色背景",
            max_tokens=4000
        )

        # 转换为 prompt
        prompt_context = aggregator.to_prompt(items)
    """

    def __init__(self, max_total_tokens: int = 8000):
        self._providers: List[IContextProvider] = []
        self._max_total_tokens = max_total_tokens

    def add_provider(self, provider: IContextProvider) -> None:
        """添加上下文提供者"""
        self._providers.append(provider)
        # 按优先级排序
        self._providers.sort(key=lambda p: p.priority)
        logger.debug(f"Added context provider: {provider.name}")

    def remove_provider(self, name: str) -> bool:
        """移除提供者"""
        for i, provider in enumerate(self._providers):
            if provider.name == name:
                self._providers.pop(i)
                return True
        return False

    async def get_context(
        self,
        query: Optional[str] = None,
        max_tokens: Optional[int] = None,
        provider_kwargs: Optional[Dict[str, Dict[str, Any]]] = None,
        **kwargs
    ) -> List[ContextItem]:
        """
        获取聚合上下文

        Args:
            query: 查询字符串
            max_tokens: 最大令牌数（None 表示无限制）
            provider_kwargs: 各提供者的额外参数
            **kwargs: 传递给所有提供者的参数

        Returns:
            按优先级排序的上下文项列表
        """
        all_items: List[ContextItem] = []
        provider_kwargs = provider_kwargs or {}

        # 从各提供者收集上下文
        for provider in self._providers:
            try:
                extra_kwargs = provider_kwargs.get(provider.name, {})
                items = await provider.get_context(query=query, **kwargs, **extra_kwargs)
                all_items.extend(items)
            except Exception as e:
                logger.error(f"Provider {provider.name} failed: {e}")

        # 按优先级排序
        all_items.sort(key=lambda x: x.priority)

        # 应用令牌预算
        if max_tokens is not None:
            all_items = self._apply_token_budget(all_items, max_tokens)

        return all_items

    def _apply_token_budget(
        self,
        items: List[ContextItem],
        max_tokens: int
    ) -> List[ContextItem]:
        """应用令牌预算，保留高优先级项目"""
        result: List[ContextItem] = []
        total_tokens = 0

        for item in items:
            if total_tokens + item.token_estimate <= max_tokens:
                result.append(item)
                total_tokens += item.token_estimate
            elif item.priority <= ContextPriority.HIGH:
                # 高优先级项目总是包含
                result.append(item)
                total_tokens += item.token_estimate
                logger.warning(f"Token budget exceeded for critical item: {item.key}")

        return result

    def to_prompt(self, items: List[ContextItem]) -> str:
        """将上下文项转换为 prompt 字符串"""
        if not items:
            return ""

        segments = [item.to_prompt_segment() for item in items]
        return "\n\n".join(segments)

    def list_providers(self) -> List[str]:
        """列出所有提供者名称"""
        return [p.name for p in self._providers]


# ============ 便捷函数 ============

def get_default_aggregator(
    memory_engine: Optional[Any] = None,
    skill_loader: Optional[Any] = None,
    project_root: Optional[str] = None
) -> ContextAggregator:
    """获取预配置的默认上下文聚合器"""
    aggregator = ContextAggregator()

    if memory_engine:
        aggregator.add_provider(MemoryContextProvider(memory_engine))

    if skill_loader:
        aggregator.add_provider(SkillContextProvider(skill_loader))

    aggregator.add_provider(ProjectContextProvider(project_root))

    return aggregator

File: src-ts/context/providers.ts
/**
 * Context Providers - Dynamic context injection system
 *
 * Aggregates context from multiple sources: memory, skills, project config.
 */

import { readFileSync, existsSync, readdirSync, statSync } from 'fs';
import { join, resolve } from 'path';
import { homedir } from 'os';
import { createLogger } from '../logger/index.js';

const log = createLogger('context-providers');

// ============================================================
// Enums & Data Classes
// ============================================================

export enum ContextPriority {
  CRITICAL = 0,
  HIGH = 10,
  NORMAL = 50,
  LOW = 100,
  OPTIONAL = 200,
}

export interface ContextItem {
  key: string;
  value: unknown;
  source: string;
  priority: ContextPriority;
  metadata: Record<string, unknown>;
  tokenEstimate: number;
}

export function createContextItem(params: {
  key: string;
  value: unknown;
  source: string;
  priority?: ContextPriority;
  metadata?: Record<string, unknown>;
  tokenEstimate?: number;
}): ContextItem {
  return {
    key: params.key,
    value: params.value,
    source: params.source,
    priority: params.priority ?? ContextPriority.NORMAL,
    metadata: params.metadata ?? {},
    tokenEstimate: params.tokenEstimate ?? 0,
  };
}

const CHINESE_CHAR_RE = /[\u4e00-\u9fff]/g;

function estimateTokens(text: string): number {
  if (!text) return 0;
  const chineseChars = (text.match(CHINESE_CHAR_RE) ?? []).length;
  const otherChars = text.length - chineseChars;
  return Math.floor(chineseChars / 1.5 + otherChars / 4);
}

function toPromptSegment(item: ContextItem): string {
  let content: string;
  if (typeof item.value === 'object' && item.value !== null) {
    content = JSON.stringify(item.value, null, 2);
  } else {
    content = String(item.value);
  }
  return `[${item.key}]\n${content}\n[/${item.key}]`;
}

// ============================================================
// Provider Interface
// ============================================================

export interface IContextProvider {
  readonly name: string;
  readonly priority: ContextPriority;
  getContext(query?: string | null, ...args: unknown[]): Promise<ContextItem[]>;
}

// ============================================================
// Base Provider
// ============================================================

export abstract class BaseContextProvider implements IContextProvider {
  readonly name: string;
  readonly priority: ContextPriority;

  constructor(name: string, priority: ContextPriority = ContextPriority.NORMAL) {
    this.name = name;
    this.priority = priority;
  }

  abstract getContext(query?: string | null, ...args: unknown[]): Promise<ContextItem[]>;

  protected _estimateTokens(text: string): number {
    return estimateTokens(text);
  }
}

// ============================================================
// Memory Context Provider
// ============================================================

export class MemoryContextProvider extends BaseContextProvider {
  private _memoryEngine: unknown;
  private _maxItems: number;
  private _relevanceThreshold: number;

  constructor(
    memoryEngine?: unknown,
    maxItems = 10,
    relevanceThreshold = 0.5,
  ) {
    super('memory', ContextPriority.HIGH);
    this._memoryEngine = memoryEngine ?? null;
    this._maxItems = maxItems;
    this._relevanceThreshold = relevanceThreshold;
  }

  async getContext(query?: string | null, ...args: unknown[]): Promise<ContextItem[]> {
    if (!this._memoryEngine) return [];

    const items: ContextItem[] = [];
    const sessionId = (args[0] as string) ?? null;

    try {
      const engine = this._memoryEngine as Record<string, unknown>;
      if (typeof engine.search === 'function') {
        const results = await (engine.search as (q: string, opts?: Record<string, unknown>) => Promise<unknown[]>)(
          query ?? '',
          { limit: this._maxItems, sessionId },
        );

        for (const result of results) {
          const r = result as Record<string, unknown>;
          if (typeof r.score === 'number' && r.score < this._relevanceThreshold) continue;

          const content = (typeof r.content === 'string' ? r.content : String(r)) as string;
          items.push(createContextItem({
            key: `memory_${items.length}`,
            value: content,
            source: this.name,
            priority: this.priority,
            metadata: {
              score: (r.score as number) ?? 1.0,
              type: (r.memory_type as string) ?? 'unknown',
            },
            tokenEstimate: this._estimateTokens(content),
          }));
        }
      }
    } catch (e) {
      log.error(`Failed to get memory context`, { error: String(e) });
    }

    return items;
  }
}

// ============================================================
// Skill Context Provider
// ============================================================

export class SkillContextProvider extends BaseContextProvider {
  private _skillLoader: unknown;
  private _maxSkillLength: number;

  constructor(skillLoader?: unknown, maxSkillLength = 4000) {
    super('skill', ContextPriority.NORMAL);
    this._skillLoader = skillLoader ?? null;
    this._maxSkillLength = maxSkillLength;
  }

  async getContext(query?: string | null, ...args: unknown[]): Promise<ContextItem[]> {
    if (!this._skillLoader) return [];

    const items: ContextItem[] = [];
    const skillIds = (args[0] as string[]) ?? [];
    const includeSummary = (args[1] as boolean) ?? true;
    const loader = this._skillLoader as Record<string, unknown>;

    try {
      const uniqueIds: string[] = [];
      const seen = new Set<string>();
      for (const id of skillIds) {
        if (!seen.has(id)) { seen.add(id); uniqueIds.push(id); }
      }

      for (const skillId of uniqueIds) {
        try {
          let content = (loader.load as (n: string) => string)(skillId);
          if (content.length > this._maxSkillLength) {
            content = content.slice(0, this._maxSkillLength) + '\n... (truncated)';
          }
          items.push(createContextItem({
            key: `skill_${skillId}`,
            value: content,
            source: this.name,
            priority: this.priority,
            metadata: { skill_id: skillId },
            tokenEstimate: this._estimateTokens(content),
          }));
        } catch { /* skill not found */ }
      }

      if (includeSummary && typeof loader.getSummary === 'function') {
        const summary = (loader.getSummary as () => string)();
        items.push(createContextItem({
          key: 'available_skills',
          value: summary,
          source: this.name,
          priority: ContextPriority.LOW,
          tokenEstimate: this._estimateTokens(summary),
        }));
      }
    } catch (e) {
      log.error(`Failed to get skill context`, { error: String(e) });
    }

    return items;
  }
}

// ============================================================
// Project Context Provider
// ============================================================

export class ProjectContextProvider extends BaseContextProvider {
  private _projectRoot: string;
  private _nikoDir: string;

  constructor(projectRoot?: string | null) {
    super('project', ContextPriority.HIGH);
    this._projectRoot = projectRoot ?? process.cwd();
    this._nikoDir = join(this._projectRoot, '.niko');
  }

  async getContext(query?: string | null, ...args: unknown[]): Promise<ContextItem[]> {
    const items: ContextItem[] = [];

    if (!existsSync(this._nikoDir)) return items;

    try {
      const includeCharacters = (args[0] as boolean) ?? true;
      const includeWorld = (args[1] as boolean) ?? true;
      const includeOutline = (args[2] as boolean) ?? true;

      const configFile = join(this._nikoDir, 'config.json');
      if (existsSync(configFile)) {
        const config = this._readJsonFile(configFile);
        items.push(createContextItem({
          key: 'project_config',
          value: config,
          source: this.name,
          priority: ContextPriority.HIGH,
          tokenEstimate: this._estimateTokens(JSON.stringify(config)),
        }));
      }

      if (includeCharacters) {
        const charsDir = join(this._nikoDir, 'characters');
        if (existsSync(charsDir)) {
          const characters: unknown[] = [];
          for (const file of readdirSync(charsDir)) {
            if (!file.endsWith('.json')) continue;
            try { characters.push(this._readJsonFile(join(charsDir, file))); } catch { /* skip */ }
          }
          if (characters.length > 0) {
            items.push(createContextItem({
              key: 'characters',
              value: characters,
              source: this.name,
              priority: ContextPriority.NORMAL,
              metadata: { count: characters.length },
              tokenEstimate: this._estimateTokens(JSON.stringify(characters)),
            }));
          }
        }
      }

      if (includeWorld) {
        const worldFile = join(this._nikoDir, 'world.json');
        if (existsSync(worldFile)) {
          const world = this._readJsonFile(worldFile);
          items.push(createContextItem({
            key: 'world',
            value: world,
            source: this.name,
            priority: ContextPriority.NORMAL,
            tokenEstimate: this._estimateTokens(JSON.stringify(world)),
          }));
        }
      }

      if (includeOutline) {
        const outlineFile = join(this._nikoDir, 'outline.json');
        if (existsSync(outlineFile)) {
          const outline = this._readJsonFile(outlineFile);
          items.push(createContextItem({
            key: 'outline',
            value: outline,
            source: this.name,
            priority: ContextPriority.HIGH,
            tokenEstimate: this._estimateTokens(JSON.stringify(outline)),
          }));
        }
      }
    } catch (e) {
      log.error(`Failed to get project context`, { error: String(e) });
    }

    return items;
  }

  private _readJsonFile(filePath: string): unknown {
    return JSON.parse(readFileSync(filePath, 'utf-8'));
  }
}

// ============================================================
// Context Aggregator
// ============================================================

export class ContextAggregator {
  private _providers: IContextProvider[] = [];
  private _maxTotalTokens: number;

  constructor(maxTotalTokens = 8000) {
    this._maxTotalTokens = maxTotalTokens;
  }

  addProvider(provider: IContextProvider): void {
    this._providers.push(provider);
    this._providers.sort((a, b) => a.priority - b.priority);
  }

  removeProvider(name: string): boolean {
    const idx = this._providers.findIndex(p => p.name === name);
    if (idx >= 0) { this._providers.splice(idx, 1); return true; }
    return false;
  }

  async getContext(
    query?: string | null,
    maxTokens?: number | null,
    providerKwargs?: Record<string, Record<string, unknown>>,
    ...kwargs: unknown[]
  ): Promise<ContextItem[]> {
    const allItems: ContextItem[] = [];
    const kwPerProvider = providerKwargs ?? {};

    const results = await Promise.allSettled(
      this._providers.map(provider => {
        const extra = kwPerProvider[provider.name] ?? {};
        return provider.getContext(query, ...kwargs, ...Object.values(extra));
      }),
    );

    for (let i = 0; i < results.length; i++) {
      const result = results[i];
      if (result.status === 'fulfilled') allItems.push(...result.value);
    }

    allItems.sort((a, b) => a.priority - b.priority);

    const limit = maxTokens ?? null;
    return limit !== null ? this._applyTokenBudget(allItems, limit) : allItems;
  }

  private _applyTokenBudget(items: ContextItem[], maxTokens: number): ContextItem[] {
    const result: ContextItem[] = [];
    let totalTokens = 0;

    for (const item of items) {
      if (totalTokens + item.tokenEstimate <= maxTokens) {
        result.push(item);
        totalTokens += item.tokenEstimate;
      } else if (item.priority <= ContextPriority.HIGH) {
        result.push(item);
        totalTokens += item.tokenEstimate;
      }
    }

    return result;
  }

  toPrompt(items: ContextItem[]): string {
    if (items.length === 0) return '';
    return items.map(toPromptSegment).join('\n\n');
  }

  listProviders(): string[] {
    return this._providers.map(p => p.name);
  }
}

// ============================================================
// Convenience Functions
// ============================================================

export function getDefaultAggregator(
  memoryEngine?: unknown,
  skillLoader?: unknown,
  projectRoot?: string | null,
): ContextAggregator {
  const aggregator = new ContextAggregator();
  if (memoryEngine) aggregator.addProvider(new MemoryContextProvider(memoryEngine));
  if (skillLoader) aggregator.addProvider(new SkillContextProvider(skillLoader));
  aggregator.addProvider(new ProjectContextProvider(projectRoot));
  return aggregator;
}
